superset-frontend/src/components/AlteredSliceTag/index.tsx

```diff
@@ -149,72 +180,39 @@ class AlteredSliceTag extends React.Component<
       if (['filters', 'having', 'where'].includes(fdKey)) {
         return;
       }
-      if (!this.isEqualish(ofd[fdKey], cfd[fdKey])) {
+      if (!isEqualish(ofd[fdKey], cfd[fdKey])) {
         diffs[fdKey] = { before: ofd[fdKey], after: cfd[fdKey] };
       }
     });
     return diffs;
-  }
+  }, [props.currentFormData, props.origFormData]);
 
-  isEqualish(val1: string, val2: string): boolean {
-    return isEqual(alterForComparison(val1), alterForComparison(val2));
-  }
+  useEffect(() => {
+    const diffs = getDiffs();
+    const controlsMap = getControlsForVizType(
+      props.origFormData?.viz_type,
+    ) as ControlMap;
+    setRows(getRowsFromDiffs(diffs, controlsMap));
+    setHasDiffs(!isEmpty(diffs));
+  }, [getDiffs, props]);
 
-  formatValue(
-    value: DiffItemType,
-    key: string,
-    controlsMap: ControlMap,
-  ): string | number {
-    if (value === undefined) {
-      return 'N/A';
-    }
-    if (value === null) {
-      return 'null';
-    }
-    if (
-      controlsMap[key]?.type === 'AdhocFilterControl' &&
-      Array.isArray(value)
-    ) {
-      if (!value.length) {
-        return '[]';
+  useEffect(() => {
+    const diffs = getDiffs();
+
+    const updateStateWithDiffs = () => {
+      if (isEqual(prevProps, props)) {
+        return;
       }
-      return value
-        .map((v: FilterItemType) => {
-          const filterVal =
-            v.comparator && v.comparator.constructor === Array
-              ? `[${v.comparator.join(', ')}]`
-              : v.comparator;
-          return `${v.subject} ${v.operator} ${filterVal}`;
-        })
-        .join(', ');
-    }
-    if (controlsMap[key]?.type === 'BoundsControl') {
-      return `Min: ${value[0]}, Max: ${value[1]}`;
-    }
-    if (
-      controlsMap[key]?.type === 'CollectionControl' &&
-      Array.isArray(value)
-    ) {
-      return value.map(v => safeStringify(v)).join(', ');
-    }
-    if (controlsMap[key]?.type === 'MetricsControl' && Array.isArray(value)) {
-      const formattedValue = value.map((v: FilterItemType) => v?.label ?? v);
-      return formattedValue.length ? formattedValue.join(', ') : '[]';
-    }
-    if (typeof value === 'boolean') {
-      return value ? 'true' : 'false';
-    }
-    if (Array.isArray(value)) {
-      const formattedValue = value.map((v: FilterItemType) => v?.label ?? v);
-      return formattedValue.length ? formattedValue.join(', ') : '[]';
-    }
-    if (typeof value === 'string' || typeof value === 'number') {
-      return value;
-    }
-    return safeStringify(value);
-  }
+      setRows(prevRows =>
+        getRowsFromDiffs(diffs, prevRows as unknown as ControlMap),
+      );
+      setHasDiffs(!isEmpty(diffs));
+    };
+
+    updateStateWithDiffs();
+  }, [getDiffs, props, prevProps]);
```

Comment:
How is that useEffect different than the one above? It looks like both will run when props change and both call the same setters

Reply:
Also, the logic in this useEffect looks a bit different than in original code, where we call `this.getRowsFromDiffs(diffs, prevState.controlsMap)` instead of `getRowsFromDiffs(diffs, prevRows)`. Is that intentional?

Reply:
this is an oversight from trying to shoehorn all the changes from the AI into the file, the second seems useEffect like a less helpful approach so I'll delete it